File: battle-royale/app/src/lib/weapon.ts
import { AnchorWallet } from '@solana/wallet-adapter-react';
import { PublicKey, TransactionInstruction } from '@solana/web3.js';
import { anchorProgram, convertWeaponTypeToAnchorPayload } from './anchor';

const TABLE_NAME = 'weapon';

export enum WeaponType {
    Dagger,
    Hammer,
    Pistol,
    Sword,
}

export async function getWeapon(wallet: AnchorWallet, id: number) {
    const program = anchorProgram(wallet);
    const [weaponPublicKey, _] = await PublicKey.findProgramAddress(
        [Buffer.from(TABLE_NAME), Buffer.from(Uint8Array.of(id))],
        program.programId,
    );
    return await program.account.weapon.fetch(weaponPublicKey);
}

export async function getAllWeapons(wallet: AnchorWallet) {
    return (await anchorProgram(wallet).account.weapon.all()).map(o => { 
        return { publicKey: o.publicKey, ...o.account }
    });
}

export async function createWeaponInstruction(
    wallet: AnchorWallet,
    id: number,
    icon: string,
    weaponName: string,
    weaponType: WeaponType,
    damage: number,
): Promise<TransactionInstruction> {
    const program = anchorProgram(wallet);
    const [weaponPublicKey, _] = await PublicKey.findProgramAddress(
        [Buffer.from(TABLE_NAME), Buffer.from(Uint8Array.of(id))],
        program.programId,
    );
    return program.methods.createWeapon(
        id, icon, weaponName, convertWeaponTypeToAnchorPayload(weaponType), damage
    )
        .accounts({
            weapon: weaponPublicKey,
            payer: wallet.publicKey,
        })
        .instruction();
}

export async function updateWeaponInstruction(
    wallet: AnchorWallet,
    id: number,
    icon: string,
    weaponName: string,
    weaponType: WeaponType,
    damage: number,
): Promise<TransactionInstruction> {
    const program = anchorProgram(wallet);
    const [weaponPublicKey, _] = await PublicKey.findProgramAddress(
        [Buffer.from(TABLE_NAME), Buffer.from(Uint8Array.of(id))],
        program.programId,
    );
    return program.methods.updateWeapon(
        id, icon, weaponName, weaponType, damage
    )
        .accounts({
            weapon: weaponPublicKey,
            payer: wallet.publicKey,
        })
        .instruction();
}

export async function deleteWeaponInstruction(
    wallet: AnchorWallet,
    id: number,
): Promise<TransactionInstruction> {
    const program = anchorProgram(wallet);
    const [weaponPublicKey, _] = await PublicKey.findProgramAddress(
        [Buffer.from(TABLE_NAME), Buffer.from(Uint8Array.of(id))],
        program.programId,
    );
    return program.methods.deleteWeapon(
        id
    )
        .accounts({
            weapon: weaponPublicKey,
            payer: wallet.publicKey,
        })
        .instruction();
}

File: battle-royale/program/programs/battle-royale/src/state/weapon.rs
use anchor_lang::prelude::*;


#[derive(AnchorDeserialize, AnchorSerialize, Clone)]
pub enum WeaponType {
    Dagger,
    Hammer,
    Pistol,
    Sword,
}

#[account]
pub struct Weapon {
    pub bump: u8,
    pub id: u8,
    pub icon: String,
    pub weapon_name: String,
    pub weapon_type: WeaponType,
    pub damage: u32,
}


impl Weapon {

    pub const INNER_DATA_SIZE: usize = 1    // .discriminator
        + 1     // bump
        + 1     // id
        + 24    // icon
        + 24    // weapon_name
        + 1     // weapon_type
        + 4;    // damage

    pub const TABLE_NAME: &'static str = "weapon";

    pub fn new(
        bump: u8,
        id: u8,
        icon: String,
        weapon_name: String,
        weapon_type: WeaponType,
        damage: u32,
    ) -> Self {
        Weapon {
            bump,
            id,
            icon,
            weapon_name,
            weapon_type,
            damage,
        }
    }
}